state/tx_options.go

```diff
@@ -0,0 +1,168 @@
+
+func (options *TxOptions) AccountKey() string { return options.accountKey }
+
+func (options *TxOptions) FeeGranterAddress() string { return options.feeGranterAddress }
+
+type jsonTxOptions struct {
+	GasPrice          float64 `json:"gas_price,omitempty"`
+	IsGasPriceSet     bool    `json:"is_gas_price_set,omitempty"`
+	Gas               uint64  `json:"gas,omitempty"`
+	AccountKey        string  `json:"account_key,omitempty"`
+	FeeGranterAddress string  `json:"fee_granter_address,omitempty"`
+}
+
+func (options *TxOptions) MarshalJSON() ([]byte, error) {
+	jsonOpts := &jsonTxOptions{
+		GasPrice:          options.gasPrice,
+		IsGasPriceSet:     options.isGasPriceSet,
+		Gas:               options.gas,
+		AccountKey:        options.accountKey,
+		FeeGranterAddress: options.feeGranterAddress,
+	}
+	return json.Marshal(jsonOpts)
+}
+
+func (options *TxOptions) UnmarshalJSON(data []byte) error {
+	var jsonOpts jsonTxOptions
+	err := json.Unmarshal(data, &jsonOpts)
+	if err != nil {
+		return fmt.Errorf("unmarshalling TxOptions: %w", err)
+	}
+
+	options.gasPrice = jsonOpts.GasPrice
+	options.isGasPriceSet = jsonOpts.IsGasPriceSet
+	options.gas = jsonOpts.Gas
+	options.accountKey = jsonOpts.AccountKey
+	options.feeGranterAddress = jsonOpts.FeeGranterAddress
+	return nil
+}
+
+// calculateFee calculates fee amount based on the `minGasPrice` and `Gas`.
+func calculateFee(gas uint64, gasPrice float64) int64 {
+	fee := int64(math.Ceil(gasPrice * float64(gas)))
+	return fee
+}
+
+// estimateGas estimates gas in case it has not been set.
+// NOTE: final result of the estimation will be multiplied by the `gasMultiplier`(1.1) to cover additional costs.
+func estimateGas(ctx context.Context, client *user.TxClient, msg sdktypes.Msg) (uint64, error) {
+	// set fee as 1utia helps to simulate the tx more reliably.
+	gas, err := client.EstimateGas(ctx, []sdktypes.Msg{msg}, user.SetFee(1))
+	if err != nil {
+		return 0, fmt.Errorf("estimating gas: %w", err)
+	}
+	return uint64(float64(gas) * gasMultiplier), nil
+}
+
+// estimateGasForBlobs returns a gas limit as a `user.TxOption` that can be applied to the `MsgPayForBlob` transactions.
+// NOTE: final result of the estimation will be multiplied by the `gasMultiplier`(1.1)
+// to cover additional options of the Tx.
+func estimateGasForBlobs(blobSizes []uint32) uint64 {
+	gas := apptypes.DefaultEstimateGas(blobSizes)
+	return uint64(float64(gas) * gasMultiplier)
+}
+
+func parseAccountKey(kr keyring.Keyring, accountKey string) (sdktypes.AccAddress, error) {
+	rec, err := kr.Key(accountKey)
+	if err != nil {
+		return nil, fmt.Errorf("getting account key: %w", err)
+	}
+	return rec.GetAddress()
+}
+
+func parseAccAddressFromString(addrStr string) (sdktypes.AccAddress, error) {
+	return sdktypes.AccAddressFromBech32(addrStr)
+}
+
+// Attribute is the functional option that is applied to the TxOption instance
+// to configure parameters.
+type Attribute func(options *TxOptions)
```

Comment:
Why not just Option?

Reply:
I would rename `TxOptions` then to something like `TxConfig/Params`

Reply:
let's call it `TxConfig`

Reply:
(note that's the reason to rename Attribute to more conventional Option)